replica/compaction_group.hh

```diff
@@ -244,19 +247,20 @@ public:
         return _compaction_groups;
     }
 
-    const storage_group_map& storage_groups() const noexcept {
-        return _storage_groups;
+    void with_storage_groups(std::function<void (const storage_group_map&)> f) const {
+        f(_storage_groups);
     }
-    storage_group_map& storage_groups() noexcept {
-        return _storage_groups;
+    future<> with_storage_groups_shared_lock(std::function<future<> (const storage_group_map&)> f) const {
+        auto holder = _lock.hold_write_lock();
+        co_await f(_storage_groups);
```

Comment:
hold_read_lock

Reply:
I think you need to consume the future returned by hold_..., otherwise you just grabbing the future and not waiting for lock to be taken.

Reply:
Right, sorry, changing

Reply:
no problem, it's one of the problem with extensive auto usage, you miss co_await and the compiler thinks it's fine because you didn't specify the type you expected. had you written `rwlock::holder holder = ...`, then miss of co_await would have been caught right away.

Reply:
no need to change it btw. was just thinking out loud